File: ts/src/boundary/eslint-scope/enter-block.ts
import type { AstNode } from "../../ir/primitive/ast-node.js";
import type { DiagnosticCollector } from "../../util/diagnostic.js";
import { hoistDeclarations } from "./hoisting/hoist-declarations.js";
import type { ScopeManager } from "./manager.js";
import type { NodeLike } from "./node-like.js";
import type { AnalysisVisitor } from "./visitor.js";
import type { PathEntry } from "./walk/path-entry.js";

export function enterBlock(
  node: NodeLike,
  parent: NodeLike | null,
  key: string | null,
  path: readonly PathEntry[],
  manager: ScopeManager,
  raw: string,
  diagnostics: DiagnosticCollector,
  visitor: AnalysisVisitor,
): void {
  const scope = manager.push("block", node as unknown as AstNode);
  visitor.onScope?.({ scope, parent, key, path });
  const stmts = node["body"];
  if (Array.isArray(stmts)) {
    hoistDeclarations(stmts, scope, raw, diagnostics);
  }
}

File: crates/unsnarl-boundary-eslint-scope/src/enter_block.rs
//! Push a `Block` scope for a `BlockStatement` and hoist its
//! declarations.
//!
//! Takes the typed `BlockStatement<'_>` whose
//! `body: Vec<Statement<'_>>` is structurally guaranteed by oxc.
//!
//! `visitor.on_scope(...)` is dispatched from
//! `ScopeBuildVisitor::fire_on_scope` (in `scope_build_visitor.rs`)
//! once this helper has pushed the scope, so the module is
//! responsible only for pushing and hoisting.

use oxc_ast::ast::BlockStatement;

use unsnarl_ir::ids::ScopeId;
use unsnarl_ir::primitive::AstNode;
use unsnarl_ir::scope_type::ScopeType;
use unsnarl_oxc_parity::AstType;

use crate::hoisting::hoist_declarations::hoist_declarations;
use crate::state::{push_scope, ScopeBuilderState};

pub(crate) fn enter_block(
    state: &mut ScopeBuilderState,
    block: &BlockStatement<'_>,
    raw: &str,
) -> ScopeId {
    let block_node = AstNode {
        r#type: AstType::BlockStatement,
        span: block.span,
    };
    let scope_id = push_scope(state, ScopeType::Block, block_node);
    hoist_declarations(state, &block.body, scope_id, raw);
    scope_id
}

#[cfg(test)]
#[path = "enter_block_test.rs"]
mod enter_block_test;
